### PR TITLE
Adding support for logging additional artifacts when using mfllow pytorch log and save model

### PR DESCRIPTION
**Pull request by shrinath-suresh**
## What changes are proposed in this pull request?

In the current version of mlflow pytorch log/save model, there is no mechanism available to push artifact files along with the pytorch model. The supporting files such as requirements.txt or vocabulary files would be really helpful in case of deployments or new environment setups. Hence, adding a new parameter artifacts to mlflow pytorch log and save model which would help user to log supporting files into mlflow.

## How is this patch tested?

Tested the code by logging artifact files along with log and save model. And also ensured that the existing scenario is working as expected(saving and logging model without artifact parameter). Added unit test cases as well.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Users can log additional artifact files such as requirements, vocabulary files along with the pytorch model when using `mlflow.pytorch.log_model` or `mfllow.pytorch.save_model` by adding them in `artifacts` parameter.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/pytorch/__init__.py [hunk 1]
```diff
@@ -325,6 +366,23 @@ def save_model(
...
+            saved_artifacts_dir_subpath = "artifacts"
+            for artifact_name, artifact_uri in artifacts.items():
+                tmp_artifact_path = _download_artifact_from_uri(
+                    artifact_uri=artifact_uri, output_path=tmp_artifacts_dir.path()
+                )
+                tmp_artifacts_config[artifact_name] = tmp_artifact_path
+
+            shutil.move(
+                tmp_artifacts_dir.path(),
+                os.path.join(path, saved_artifacts_dir_subpath),
```
Comment (shrinath-suresh) [inline mlflow/pytorch/__init__.py:384]: will change it to posix path and upload a newer version soon

File: tests/pytorch/test_pytorch_model_export.py [hunk 2]
```diff
@@ -811,3 +812,86 @@ def test_sagemaker_docker_model_scoring_with_sequential_model_and_default_conda_
...
+        assert os.path.isfile(requirements_file) is True
+        with open(requirements_file) as fp:
+            content = fp.read()
+
+        assert "pytest" in content
+
+
+def test_log_model_invalid_path(create_artifact, sequential_model):
+    artifact_file_path = create_artifact
+    print("Artifact Path {}".format(artifact_file_path))
```
Comment (harupy) [inline tests/pytorch/test_pytorch_model_export.py:873]: @shrinath-suresh Can you remove this line?

File: tests/pytorch/test_pytorch_model_export.py [hunk 3]
```diff
@@ -811,3 +812,86 @@ def test_sagemaker_docker_model_scoring_with_sequential_model_and_default_conda_
...
+                "requirements.txt": str(artifact_file_path).replace(
+                    "requirements", "requirements1"
+                )
+            },
+        )
+
+
+def test_log_model_invalid_type(create_artifact, sequential_model):
+    artifact_file_path = create_artifact
+    print("Artifact Path {}".format(artifact_file_path))
```
Comment (harupy) [inline tests/pytorch/test_pytorch_model_export.py:889]:
```suggestion
```

File: tests/pytorch/test_pytorch_model_export.py (hunk 2 above)
Comment (shrinath-suresh) [reply tests/pytorch/test_pytorch_model_export.py]: Have removed and uploaded a new version

**Comment by shrinath-suresh** [comment]
Something went wrong when i signed off the previous commits :-( . In the DCO error log command, it was showing 8 commits, However, it was actually only 3 which i had changed.

**Comment by shrinath-suresh** [comment]
@harupy Can i delete this branch and PR and raise a new one ? Finding it hard to revert all the changes mentioned above

**Comment by harupy** [comment]
@shrinath-suresh Yes you can do that but how did you signoff your commits?

**Comment by shrinath-suresh** [comment]
> @shrinath-suresh Yes you can do that but how did you signoff your commits?

DCO failed when i removed and uploaded changes. I went to details and then it showed

```
git rebase HEAD~8 --signoff
git push --force-with-lease origin pytorch-remove-cloudpickle

```

After running this i realised that there were in between commits by others and so many files were added. Is there any better way to handler this signoff previous commits ?  - (i will keep in mind to add sign off while committing the code :-) )

**Comment by harupy** [comment]
> DCO failed when i removed and uploaded changes.

@shrinath-suresh What command did you use when you removed and uploaded changes?

**Comment by shrinath-suresh** [comment]
> > DCO failed when i removed and uploaded changes.
> 
> @shrinath-suresh **before running the DCO command**, what commands did you run?

i did `git pull origin master` and then tried to sign off using the commands i mentioned. Looks like my master branch has some commits which arent signed off. Even if i clone a new branch, DCO fails since some commits are not signed off. Can you please guide me on how to sign the previous commits ?

**Comment by harupy** [comment]
Your master branch seems to have commits that mlflow's master brach doesn't have...

**Comment by shrinath-suresh** [comment]
> Your master branch seems to have commits that mlflow's master brach doesn't have...
> 
> https://github.com/shrinath-suresh/mlflow/commits/master

Yes.. At first , I did the save artifacts commit in master.. Since, multiple PRs were needed, i reverted the commits from master. Now, when i create a branch out of master, those commits (which are reverted and  not signed off) are getting failed in DCO.

**Comment by harupy** [comment]
I think you'll get in trouble with your master branch.

**Comment by harupy** [comment]
We provide a pre-commit git hook that validates that commits are signed-off.

```python
# Run this command in the repo root.
git config core.hooksPath hooks
```

**Comment by shrinath-suresh** [comment]
> @shrinath-suresh I think you'll get in trouble again with your current master branch. Can you reset it to upstream ([ref](https://ardalis.com/syncing-a-fork-of-a-github-repository-with-upstream/)) ?

Thank you very much . The link helped very much to hard reset the master. i have shared a new pull request now.

**Comment by shrinath-suresh** [comment]
Create a new pull request - https://github.com/mlflow/mlflow/pull/3436 

@harupy Can you please review when you get time?

**Pull request closed without merging**